### ISSUE TITLE
can't transform a range of boost::filesystem::directory_iterator

### ISSUE THREAD

**Issue by etam** [ericniebler/range-v3#602]
This works:
```
#include <iostream>
#include <boost/filesystem.hpp>
#include <range/v3/all.hpp>

namespace fs = boost::filesystem;
namespace r = ranges;

int main()
{
    const auto path = fs::path{"."};
    const auto ls = r::make_iterator_range(fs::directory_iterator{path},
                                           fs::directory_iterator{});
    r::for_each(ls, [](auto&& i) {
            std::cout << i.path().filename().string() << '\n';
        });
}
```

This does not compile:
```
#include <iostream>
#include <boost/filesystem.hpp>
#include <range/v3/all.hpp>

namespace fs = boost::filesystem;
namespace r = ranges;
namespace rv = r::view;

int main()
{
    const auto path = fs::path{"."};
    const auto ls = r::make_iterator_range(fs::directory_iterator{path},
                                           fs::directory_iterator{});
    const auto names = ls | rv::transform([](auto&& i) {
            return i.path().filename().string();
        });
    r::for_each(names, [](auto&& name) {
            std::cout << name << '\n';
        });
}
```
gcc 6.2.1 output: http://susepaste.org/95072834
clang 3.8.0 output: http://susepaste.org/6936650
ranges git commit 5941ef6

**Comment by CaseyCarter** [comment]
It would appear that `directory_iterator` is lying about having a viable `-` operator. I'm unable to repro this on my Yakkety setup with boost trunk. What boost version are you using?

**Comment by etam** [comment]
I have boost 1.54.0
And it seems to be fixed in 1.63 https://svn.boost.org/trac/boost/ticket/12578

Sorry, for making trouble.
